packages/runtime/src/runtime.ts

```diff
@@ -324,19 +347,160 @@ export class Runtime {
+  }
+
   /**
-   * Returns Asset Definitions
+   * https://developer.algorand.org/docs/features/asa/#modifying-an-asset
+   * Modifies asset fields
+   * @param sender sender address
    * @param assetId Asset Index
+   * @param fields Asset modifying fields
+   * @param payFlags Transaction Parameters
    */
-  getAssetDef (assetId: number): AssetDef {
-    const addr = this.store.assetDefs.get(assetId);
-    if (addr === undefined) {
-      throw new TealError(ERRORS.ASA.ASSET_NOT_FOUND, { assetId: assetId });
+  modifyAsset (sender: string, assetId: number, fields: AssetModFields, payFlags: TxParams): void {
+    this.mkAssetConfigTx(sender, assetId, fields, payFlags);
+    const creatorAcc = this.getAssetAccount(assetId);
+    creatorAcc.modifyAsset(assetId, fields);
+  }
+
+  // creates asset freeze transaction object.
+  mkAssetFreezeTx (
+    sender: string, assetId: number, target: string, state: boolean, flags: TxParams): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetFreezeTxnWithSuggestedParams(
+      sender,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      target,
+      state,
+      mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetId);
+    if (asset.freeze !== sender) {
+      throw new Error("Only Freeze account can freeze asset");
     }
+  }
 
-    const creatorAcc = this.assertAccountDefined(addr, this.store.accounts.get(addr));
-    const assetDef = creatorAcc.getAssetDef(assetId);
-    return this.assertAssetDefined(assetId, assetDef);
+  /**
+   * https://developer.algorand.org/docs/features/asa/#freezing-an-asset
+   * Freezes assets for a target account
+   * @param sender sender address
+   * @param assetId asset index
+   * @param freezeTarget target account
+   * @param freezeState target state
+   * @param payFlags transaction parameters
+   */
+  freezeAsset (
+    sender: string, assetId: number, freezeTarget: string,
+    freezeState: boolean, payFlags: TxParams
+  ): void {
+    this.mkAssetFreezeTx(sender, assetId, freezeTarget, freezeState, payFlags);
+    const acc = this.assertAccountDefined(freezeTarget, this.store.accounts.get(freezeTarget));
+    acc.freezeAsset(assetId, freezeState);
+  }
+
+  // creates asset revoke transaction object.
+  mkAssetRevokeTx (
+    sender: string, recipient: string, assetId: number,
+    revocationTarget: string, amount: number, flags: TxParams, assetID: number
+  ): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetTransferTxnWithSuggestedParams(
+      sender,
+      recipient,
+      flags.closeRemainderTo,
+      revocationTarget,
+      amount,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetID);
+    if (asset.clawback !== sender) {
+      throw new Error("Only Clawback account can revoke assets");
```

Comment:
```suggestion
      throw new Error(`Only Clawback account(${asset.clawback}) can revoke assets`);
```

Reply:
Also, should we move this check before `makeAssetTransferTxnWithSuggestedParams`?

Reply:
no, because that's not the flow.
first we create the tx and then send it to network

Reply:
Yes, but this entire function constitues mkTx right? I was thinking to avoid calling SDK constructor, if we can throw this error beforeHand.

Reply:
can you please elaborate ?

Reply:
This function is `mkTx` function, i am assuming that line `437` is our validation during createTx (after SDK). So we could do this validation before calling sdk (so if error is thrown we don't need to call an external module(`algosdk`) constructor). Ignore if that's not case.

Reply:
we cannot do that because information for asset.clawback is stored in network, therfore first we create tx and then do what network does